File: Problems/3. Longest Substring Without Repeating Characters/longest-substring-without-repeating-characters.rb
#Solution 1: 110ms 211 MB
def length_of_longest_substring(s)
  return 1 if s.length == 1

  longestSubstring = ""
  tempLongestSubstring = ""

  s.each_char do |char|
    indexOfChar = tempLongestSubstring.index(char)
    if indexOfChar != nil
      if tempLongestSubstring.length > longestSubstring.length
        longestSubstring = tempLongestSubstring 
      end
      tempLongestSubstring = tempLongestSubstring[indexOfChar + 1..-1]
    end
    tempLongestSubstring += char
  end

  if tempLongestSubstring.length > longestSubstring.length
    longestSubstring = tempLongestSubstring 
  end

  return longestSubstring.length
end

File: Problems/3. Longest Substring Without Repeating Characters/longest-substring-without-repeating-characters.ts
//Solution 1: 75ms 48.4 MB
function lengthOfLongestSubstring(s: string): number {
  let longestSubstring = "";
  let tempLongestSubstring = "";

  for (let i = 0; i < s.length; i++) {
    const indexOfChar = tempLongestSubstring.indexOf(s[i]);
    if (indexOfChar !== -1) {
      if (tempLongestSubstring.length > longestSubstring.length) {
        longestSubstring = tempLongestSubstring;
      }
      tempLongestSubstring = tempLongestSubstring.substring(indexOfChar + 1);
    }

    tempLongestSubstring += s[i];
  }

  if (tempLongestSubstring.length > longestSubstring.length) {
    longestSubstring = tempLongestSubstring;
  }

  return longestSubstring.length;
}

//Solution 2: 68ms 48.9 MB
function lengthOfLongestSubstring(s: string): number {
  if (s.length === 1) return 1;

  let longestSubstring = "";
  let tempLongestSubstring = "";

  for (let i = 0; i < s.length; i++) {
    const indexOfChar = tempLongestSubstring.indexOf(s[i]);
    if (indexOfChar !== -1) {
      if (tempLongestSubstring.length > longestSubstring.length) {
        longestSubstring = tempLongestSubstring;
      }
      tempLongestSubstring = tempLongestSubstring.substring(indexOfChar + 1);
    }

    tempLongestSubstring += s[i];
  }

  if (tempLongestSubstring.length > longestSubstring.length) {
    longestSubstring = tempLongestSubstring;
  }

  return longestSubstring.length;
}
